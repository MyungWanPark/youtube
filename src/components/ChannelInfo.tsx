import React from 'react';
import { ChannelProps } from '../types/videoType';
import { useQuery } from '@tanstack/react-query';
import { useYoutube } from './../context/YoutubeApiContext';

type Props = {
    info: ChannelProps;
};

export default function ChannelInfo({ info }: Props) {
    const { youtube } = useYoutube();
    const { isLoading, error, data } = useQuery(['channelInfo', info.id], () => youtube.channelImageURL(info.id));
    return <div>{data && <img src={data} alt={info.title} />}</div>;
}
