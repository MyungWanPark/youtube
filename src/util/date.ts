import { format, register } from 'timeago.js';
import koLocale from 'timeago.js/lib/lang/ko.js';

register('ko', koLocale);

export function formatAgo(date: string, lang: string = 'en_US') {
    return format(date, lang);
}
