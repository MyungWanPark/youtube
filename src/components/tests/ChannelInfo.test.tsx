import { withAllContext, withRouter } from '../../test/util';
import ChannelInfo from '../ChannelInfo';
import { render, screen } from '@testing-library/react';
/**
 * Test 할 요소
 * 1. useQuery를 통한 youtube.channelImageURL을 잘 받아오는지?
 */

describe('ChannelInfo', () => {
    const youtubeClient = {
        channelImageURL: jest.fn(),
        search: jest.fn(),
        popular: jest.fn(),
        relatedVideo: jest.fn(),
    };
    afterEach(() => {
        youtubeClient.channelImageURL.mockReset();
    });

    test('renders correctly with imageURL', async () => {
        const testProps = {
            id: '1',
            title: 'title1',
        };

        youtubeClient.channelImageURL.mockImplementation(() => new Promise((resolve, reject) => resolve('/imageURL')));

        const { asFragment } = render(
            withAllContext(
                withRouter([
                    {
                        path: '/',
                        element: <ChannelInfo info={testProps} />,
                    },
                ]),
                youtubeClient
            )
        );

        const image = (await screen.findByRole('img')) as HTMLImageElement;
        expect(image.src.includes('imageURL')).toBe(true);
        expect(image.alt).toBe('title1');
        expect(asFragment()).toMatchSnapshot();
    });

    test('render without URL', async () => {
        youtubeClient.channelImageURL.mockImplementation(() => {
            throw new Error('image not found');
        });
        const testProps = {
            id: '1',
            title: 'title1',
        };

        render(
            withAllContext(
                withRouter([
                    {
                        path: '/',
                        element: <ChannelInfo info={testProps} />,
                    },
                ]),
                youtubeClient
            )
        );

        expect(screen.queryByRole('img')).toBeNull();
    });
});
