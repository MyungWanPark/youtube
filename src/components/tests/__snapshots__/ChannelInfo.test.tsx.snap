// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelInfo renders correctly 1`] = `
<DocumentFragment>
  <div
    class="my-4 mb-8 flex items-center"
  >
    <img
      alt="title1"
      class="w-10 h-10 rounded-full mr-2"
      src="/imageURL"
    />
    <p
      class="text-lg font-medium"
    >
      title1
    </p>
  </div>
</DocumentFragment>
`;

exports[`ChannelInfo renders correctly with imageURL 1`] = `
<DocumentFragment>
  <div
    class="my-4 mb-8 flex items-center"
  >
    <img
      alt="title1"
      class="w-10 h-10 rounded-full mr-2"
      src="/imageURL"
    />
    <p
      class="text-lg font-medium"
    >
      title1
    </p>
  </div>
</DocumentFragment>
`;
